Harbor DB reached connection limit after upgradation

The DB connection reaches the up limit after upgrade from v2.2.0 to v2.3.2.

It has a minutely replication policy with id 1 in the DB, and from the log, it seems that the jobservice continue sending webhook to core.

error in core

```
Oct  5 16:00:14 172.20.1.1 core[30342]: 2021-10-05T16:00:14Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445677, ignore
Oct  5 16:00:19 172.20.1.1 core[30342]: 2021-10-05T16:00:19Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445772, ignore
Oct  5 16:00:22 172.20.1.1 core[30342]: 2021-10-05T16:00:22Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445937, ignore
Oct  5 16:02:31 172.20.1.1 core[30342]: 2021-10-05T16:02:31Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445876, ignore
Oct  5 16:02:44 172.20.1.1 core[30342]: 2021-10-05T16:02:44Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445709, ignore
Oct  5 16:02:58 172.20.1.1 core[30342]: 2021-10-05T16:02:58Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445840, ignore
Oct  5 16:03:07 172.20.1.1 core[30342]: 2021-10-05T16:03:07Z [WARNING] [/pkg/scheduler/callback.go:91]: got no schedule record with ID 1 and revision < 1633445772, ignore
```

error in jobservcie

```
Scheduled(rev=1633412796)
Oct  5 17:12:30 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:30Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633399130, status=Scheduled, revision=1633399043->http://core:8080/service/notifications/jobs/replication/1, duration=41.407843152s
Oct  5 17:12:31 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:31Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633447280, status=Scheduled, revision=1633447230->http://core:8080/service/notifications/jobs/replication/1, duration=2.806736406s
Oct  5 17:12:36 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:36Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633434890, status=Scheduled, revision=1633434800->http://core:8080/service/notifications/jobs/replication/1, duration=7.847441086s
Oct  5 17:12:37 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:37Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633439690, status=Scheduled, revision=1633439676->http://core:8080/service/notifications/jobs/replication/1, duration=38.898950467s
Oct  5 17:12:38 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:38Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633428650, status=Scheduled, revision=1633428561->http://core:8080/service/notifications/jobs/replication/1, duration=19.718132439s
Oct  5 17:12:38 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:38Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633411510, status=Scheduled, revision=1633411422->http://core:8080/service/notifications/jobs/replication/1, duration=57.570950225s
Oct  5 17:12:39 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:39Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633407920, status=Scheduled, revision=1633407919->http://core:8080/service/notifications/jobs/replication/1, duration=34.279791125s
Oct  5 17:12:39 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:39Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers, evt=status change: job=0239850290bd0fce10f0a8f7@1633412900, status=Scheduled, revision=1633412796->http://core:8080/service/notifications/jobs/replication/1, duration=1.27803321s
Oct  5 17:12:39 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:39Z [ERROR] [/jobservice/runtime/bootstrap.go:152]: trigger hook event error: Post "http://core:8080/service/notifications/jobs/replication/1": net/http: timeout awaiting response headers
Oct  5 17:12:39 172.20.1.1 jobservice[21312]: 2021-10-05T17:12:39Z [INFO] [/jobservice/worker/cworker/reaper.go:194]: Reaper: fire hook again for job 0239850290bd0fce10f0a8f7@1633413270 as job status change is not ACKed: 
```

Workaround:

Remove the redis db record of jobservice for replication
Remove all of the execution/task/schdedule of replication

[jobservice.log](https://github.com/goharbor/harbor/files/7305247/jobservice.log)

--- a/src/controller/replication/execution.go
+++ b/src/controller/replication/execution.go
@@ -87,7 +87,7 @@ func NewController() Controller {
 		scheduler:  scheduler.Sched,
 		flowCtl:    flow.NewController(),
 		ormCreator: orm.Crt,
-		wp:         lib.NewWorkerPool(1024),
+		wp:         lib.NewWorkerPool(10),
 	}
 }
 
@@ -113,13 +113,15 @@ func (c *controller) Start(ctx context.Context, policy *replicationmodel.Policy,
 	if err != nil {
 		return 0, err
 	}
-	c.wp.GetWorker()
 	// start the replication flow in background
 	// as the process runs inside a goroutine, the transaction in the outer ctx
-	// may be submitted already when the process starts, so pass a new context
+	// may be submitted already when the process starts, so create an new context
 	// with orm populated to the goroutine
-	go func(ctx context.Context) {
+	go func() {
+		c.wp.GetWorker()
 		defer c.wp.ReleaseWorker()
+
+		ctx := orm.NewContext(context.Background(), c.ormCreator.Create())
 		// recover in case panic during the adapter process
 		defer func() {
 			if err := recover(); err != nil {
@@ -145,7 +147,7 @@ func (c *controller) Start(ctx context.Context, policy *replicationmodel.Policy,
 			return
 		}
 		c.markError(ctx, id, err)
-	}(orm.NewContext(context.Background(), c.ormCreator.Create()))
+	}()
 	return id, nil
 }
 

--- a/src/pkg/reg/adapter/adapter.go
+++ b/src/pkg/reg/adapter/adapter.go
@@ -26,7 +26,7 @@ import (
 
 // const definition
 const (
-	MaxConcurrency = 100
+	MaxConcurrency = 30
 )
 
 var registry = map[string]Factory{}

